services/jokeService.js

```diff
@@ -0,0 +1,118 @@
+const https = require('https');
+
+const DataBaseClient = require('../db/DataBaseClient');
+
+const { getRandomInt } = require('../utils');
+
+const fileName = 'jokes';
+const jokesDBClient = new DataBaseClient(fileName);
+
+function apiRequest(url) {
+    const options = {
+        headers: { 'Accept': 'application/json' },
+        method: 'GET'
+    };
+
+    return new Promise((resolve, reject) => {
+        https
+            .request(url, options, (response) => {
+                const body = [];
+
+                response.on('data', chunk => body.push(chunk));
+                response.on('end', () => resolve(Buffer.concat(body)
+                    .toString()));
+            })
+            .on('error', err => reject(err.message))
+            .end();
+    });
+}
+
+async function getJokes(searchParams = {}) {
+    const apiUrl = new URL('https://icanhazdadjoke.com');
+    const searchParamsKeys = Object.keys(searchParams);
+
+    let jokes = [];
+
+    if (searchParamsKeys.length) {
+        apiUrl.pathname = '/search';
+        searchParamsKeys.forEach(key => apiUrl.searchParams.set(key, searchParams[key]));
+    }
+
+    const firstPageRawData = await apiRequest(apiUrl);
+    const {
+        current_page: currentPage,
+        next_page: nextPage,
+        total_pages: totalPages,
+        results: firstPageJokes
+    } = JSON.parse(firstPageRawData);
+
+    if (currentPage < totalPages) {
+        const pagesRequests = [];
+
+        for (let i = nextPage; i <= totalPages; i += 1) {
+            apiUrl.searchParams.set('page', i);
+
+            pagesRequests.push(apiRequest(apiUrl));
+        }
+
+        const pagesRawData = await Promise.all(pagesRequests);
+        const pagesData = pagesRawData.map(data => JSON.parse(data));
+
+        jokes = pagesData.reduce((mergedJokes, { results: pageJokes }) => mergedJokes.concat(pageJokes), firstPageJokes);
+    } else {
+        jokes = firstPageJokes;
+    }
+
+    if (!jokes.length) {
+        throw new Error('No jokes were found for that search term.');
+    }
+
+    return jokes;
+}
+
+function findMostCommonJoke(jokes) {
+    if (!jokes.length) {
+        throw new Error('There is no data to show.');
+    }
+
+    // idsWithQuantity = { [id]: [number of times the id was found], ... }
+    const idsWithQuantity = jokes.reduce((collection, { id }) => {
+        if (!collection[id]) {
+            collection[id] = 0;
+        }
+        collection[id] += 1;
+
+        return collection;
+    }, {});
+
+    // sorted in descending order and got the first id
+    const mostCommonId = Object.keys(idsWithQuantity)
+        .sort((a, b) => idsWithQuantity[b] - idsWithQuantity[a])[0];
+
+    return jokes.find(({ id }) => id === mostCommonId);
+}
+
+module.exports.getJokeByTerm = async function (term) {
+    try {
+        const jokes = await getJokes({ term });
+        const randomIdx = getRandomInt(0, jokes.length);
+        const { id, joke } = jokes[randomIdx];
```

Comment:
it is strange to have property joke in joke

Reply:
@BSiarhei but there is **jokes**